src/overlay/EnvelopeHandler.h

```diff
@@ -0,0 +1,71 @@
+#pragma once
+
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#include "overlay/Peer.h"
+#include "transactions/TransactionFrame.h"
+#include "xdr/Stellar-overlay.h"
+
+namespace stellar
+{
+
+struct ItemKey;
+
+class EnvelopeHandler
```

Comment:
Some of the changes look good. I think that there are some fundamental issues in this PR though:
* you kept `PendingEnvelope` and `ReadyEnvelope` as one, but this causes all sorts of problems: dependency loops, questionable data ownership.
   * I think it would be much cleaner to move the "pending" part into overlay (like you did) but keep the "ready" part in herder.
   * Concern of overlay will be to feed (and maybe cache in the future) full envelopes (with QSet and TxSet) to herder
   * that way we don't need new layers of abstraction to just do "processSCPQueue"
* we should start thinking about getting rid of some of the broken semantics of current's implementation (that you kept from the item fetcher design)
   * right now we have recursive calls that happen when for example we close a ledger: we craft a quorum set from the local state and by feeding it into FutureEnvelope we turn an envelope into "ready state". It makes the code hard to follow, ideally we would just have overlay (later) call "ReceiveEnvelope".